src/Functions/FunctionGenerateRandomStructure.cpp

```diff
@@ -84,119 +93,147 @@ class FunctionGenerateRandomStructure : public IFunction
+        TypeIndex::Float64,
+        TypeIndex::Date,
+        TypeIndex::Date32,
+        TypeIndex::DateTime,
+        TypeIndex::String,
+        TypeIndex::FixedString,
+        TypeIndex::IPv4,
+        TypeIndex::IPv6,
+        TypeIndex::UUID,
+    };
 
-    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    template <bool allow_complex_types>
+    constexpr auto getAllTypes()
     {
-        if (arguments.size() > 2)
-            throw Exception(
-                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
-                "Number of arguments for function {} doesn't match: passed {}, expected from 0 to 2",
-                getName(), arguments.size());
+        constexpr size_t complex_types_size = complex_types.size() * allow_complex_types;
+        constexpr size_t result_size = simple_types.size() + complex_types_size;
+        std::array<TypeIndex, result_size> result;
+        size_t index = 0;
 
+        for (size_t i = 0; i != simple_types.size(); ++i, ++index)
+            result[index] = simple_types[i];
 
-        for (size_t i = 0; i != 2; ++i)
-        {
-            if (arguments.size() == i)
-                break;
+        for (size_t i = 0; i != complex_types_size; ++i, ++index)
+            result[index] = complex_types[i];
 
-            if (!isUnsignedInteger(arguments[i]) && !arguments[i]->onlyNull())
-            {
-                throw Exception(
-                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
-                    "Illegal type {} of the {} argument of function {}, expected unsigned integer or Null",
-                    i + 1,
-                    arguments[i]->getName(),
-                    getName());
-            }
-        }
+        return result;
+    }
 
-        return std::make_shared<DataTypeString>();
+    size_t generateNumberOfColumns(pcg64 & rng)
+    {
+        return rng() % MAX_NUMBER_OF_COLUMNS + 1;
     }
 
-    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    void writeLowCardinalityNestedType(pcg64 & rng, WriteBuffer & buf, bool allow_suspicious_lc_types, bool allow_nullable = true)
     {
-        size_t seed = randomSeed();
-        size_t number_of_columns = 0;
+        bool make_nullable = allow_nullable & rng() % 2;
+        if (make_nullable)
+            writeCString("Nullable(", buf);
 
-        if (!arguments.empty() && !arguments[0].column->onlyNull())
+        if (allow_suspicious_lc_types)
         {
-            number_of_columns = arguments[0].column->getUInt(0);
-            if (number_of_columns > MAX_NUMBER_OF_COLUMNS)
-                throw Exception(
-                    ErrorCodes::BAD_ARGUMENTS,
-                    "Maximum allowed number of columns is {}, got {}",
-                    MAX_NUMBER_OF_COLUMNS,
-                    number_of_columns);
+            TypeIndex type = suspicious_lc_types[rng() % map_key_types.size()];
```

Comment:
1. `% suspicious_lc_types.size()`
2. Float32, Float64, and IPv6 are allowed in LowCardinality, but not if it's map key. So, need two flavors of this function: for map key and not, I guess. Or maybe we should make maps allow these key types.

Reply:
It turned out that `Map` keys could be only `LowCardinality(FixedString(N))` and `LowCardinality(String)` (seems like we just don't support suspicious types here).

Reply:
Ah, sorry, I thought I checked it, must have mixed it up with all the other combinations I checked while reviewing.